### PR TITLE
Added  automatic module registration

### PR DESCRIPTION
**Pull request by DasDirt**
Added  automatic module registration & Converted the Module class in kotlin

### CODE REVIEW

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt [hunk 1]
```diff
@@ -4,24 +4,16 @@ import net.ccbluex.liquidbounce.LiquidBounce
...
 import net.minecraftforge.fml.relauncher.Side
 import net.minecraftforge.fml.relauncher.SideOnly
+import org.reflections.Reflections
 
 /**
  * LiquidBounce Hacked Client
  * A minecraft forge injection client using Mixin
  *
- * @author CCBlueX
+ * @author CCBlueX, DasDirt
```
Comment (NurMarvin) [inline 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt]: The author attribute should stay only CCBlueX accross all of the client.

Comment (DasDirt) [reply 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt]: youre right

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt [hunk 2]
```diff
@@ -4,25 +4,18 @@ import net.ccbluex.liquidbounce.LiquidBounce
...
 import net.minecraftforge.fml.relauncher.SideOnly
+import org.reflections.Reflections
 
 /**
  * LiquidBounce Hacked Client
  * A minecraft forge injection client using Mixin
  *
  * @author CCBlueX
  * @game Minecraft
+ * Edited by DasDirt
```
Comment (superblaubeere27) [inline 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt]: Das gehört zu @author

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt [hunk 3]
```diff
@@ -36,157 +29,12 @@ object ModuleManager : Listenable {
...
-                TNTESP::class.java,
-                ComponentOnHover::class.java,
-                KeepAlive::class.java,
-                ResourcePackSpoof::class.java,
-                NoSlowBreak::class.java
-        )
-
+        for (aClass in Reflections(ModuleManager::class.java.getPackage().toString().replace("package ", "")).getTypesAnnotatedWith(ModuleInfo::class.java)) {
+            registerModule(aClass.newInstance() as Module)
+        }
```
Comment (superblaubeere27) [inline 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt]: Kann Probleme beim Remapping geben, addet Module, die Experimental seien sollen, muss man gucken, ob man das will

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/command/CommandManager.kt [hunk 4]
```diff
@@ -1,9 +1,13 @@
 package net.ccbluex.liquidbounce.features.command
 
 import net.ccbluex.liquidbounce.features.command.commands.*
+import net.ccbluex.liquidbounce.features.module.Module
+import net.ccbluex.liquidbounce.features.module.ModuleInfo
+import net.ccbluex.liquidbounce.features.module.ModuleManager
 import net.ccbluex.liquidbounce.utils.ClientUtils
 import net.minecraftforge.fml.relauncher.Side
 import net.minecraftforge.fml.relauncher.SideOnly
+import org.reflections.Reflections
```
Comment (superblaubeere27) [inline 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/command/CommandManager.kt]: Unnötig

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt (hunk 2 above)
Comment (DasDirt) [reply 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt]: I've moved it to @author

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt (hunk 3 above)
Comment (DasDirt) [reply 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/ModuleManager.kt]: Ich hatte bisher auch bei persönlichen projecten keine Probleme damit. Damit experimentelle Module nicht geadded werden könnte ich einen boolean in moduleinfo einbauen

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/command/CommandManager.kt (hunk 4 above)
Comment (DasDirt) [reply 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/command/CommandManager.kt]: a new dependency isn't really need, but it makes it easier.

**Comment by Marco09MC** [comment]
I had to revert your automatic module registration.
https://github.com/CCBlueX/LiquidBounce/commit/a9530f4cb57927fdbf629160b8c441d70701699a

It's using too much libraries for such a little task like adding modules.

**Comment by DasDirt** [comment]
I'll rewrite that later so that you don't need a lib for it

**Comment by cookiedragon234** [comment]
I think this could be done better without a library.

One possibility would be to scan the Launch.classloader.resourceCache for resources that end with a `.class` and start with `net.ccbluex.liquidbounce.features.module.modules.`. Peform a Class.forClass using this same classloader in order to force it to be defined. Then initialise it by fetching the constructor instance.

I will try this out on my own client and post a code snippet if I can get it to work ;)

Also on an unrelated note is there a discord for liquid bounce?

**Comment by SenkJu** [comment]
I don't like the idea of automatically registering modules. It just seems kind of hacky to me. 

We have a Discord. https://discord.gg/KzDEVxW

**Comment by cookiedragon234** [comment]
Done, using guava which is already included within forge

```Java
try
{
	for (ClassPath.ClassInfo classInfo : ClassPath.from(Launch.classLoader).getAllClasses())
	{
		if (classInfo.getName().startsWith("cookiedragon.luchadora."))
		{
			try
			{
				Class<?> aClass = classInfo.load();
				if (!Modifier.isAbstract(aClass.getModifiers()) && AbstractModule.class.isAssignableFrom(aClass))
				{
					System.out.println("Found " + classInfo.getName());
					for (Constructor<?> constructor : aClass.getConstructors())
					{
						if (constructor.getParameterCount() == 0)
						{
							AbstractModule instance = (AbstractModule)constructor.newInstance();
							modules.put(instance, instance.getValues());
							break;
						}
					}
				}
			}
			catch (Exception e)
			{
				e.printStackTrace();
			}
		}
	}
}
catch (Exception e)
{
	throw new RuntimeException("Failure loading modules", e);
}
```

This is the implementation for my project, but it shouldnt be too hard for you to adapt it for yours

**Pull request merged**